libraries/GCS_MAVLink/GCS.h

```diff
@@ -613,6 +614,84 @@ class GCS_MAVLINK
 
     uint8_t send_parameter_async_replies();
 
+#if HAVE_FILESYSTEM_SUPPORT
+
+    enum class FTP_OP : uint8_t {
+        None = 0,
+        TerminateSession = 1,
+        ResetSessions = 2,
+        ListDirectory = 3,
+        OpenFileRO = 4,
+        ReadFile = 5,
+        CreateFile = 6,
+        WriteFile = 7,
+        RemoveFile = 8,
+        CreateDirectory = 9,
+        RemoveDirectory = 10,
+        OpenFileWO = 11,
+        TruncateFile = 12,
+        Rename = 13,
+        CalcFileCRC32 = 14,
+        BurstReadFile = 15,
+        Ack = 128,
+        Nack = 129,
+    };
+
+    enum class FTP_ERROR : uint8_t {
+        None = 0,
+        Fail = 1,
+        FailErrno = 2,
+        InvalidDataSize = 3,
+        InvalidSession = 4,
+        NoSessionsAvailable = 5,
+        EOF_ = 6, // FIXME: Maybe don't send this in
+        UnknownCommand = 7,
+        FileExists = 8,
+        FileProtected = 9,
+        FileNotFound = 10,
+    };
+
+    struct pending_ftp {
+        uint32_t offset;
+        mavlink_channel_t chan;        
+        uint16_t seq_number;
+        FTP_OP opcode;
+        FTP_OP req_opcode;
+        bool  burst_complete;
+        uint8_t  size;
+        uint8_t  session;
+        uint8_t sysid;
+        uint8_t compid;
+        uint8_t  data[239];
+    };
+
+    enum class FTP_FILE_MODE {
+        Read,
+        Write,
+    };
+
+    struct ftp_state {
+        ObjectBuffer<pending_ftp> *requests;
+        ObjectBuffer<pending_ftp> *replies;
+
+        // session specific info, currently only support a single session over all links
+        int fd = -1;
+        FTP_FILE_MODE mode; // work around AP_FS not supporting file modes
```

Comment:
```suggestion
        FTP_FILE_MODE mode; // work around AP_Filesystem not supporting file modes
```

Reply:
I've made the change. As a note I actually prefer AP_FS just because it's a lot shorter to type, and the singleton is actually accessed as `AP::FS()`